interpreter.go

```diff
@@ -20,159 +20,70 @@ import (
 	"bytes"
 	"fmt"
 	"math"
+	"path"
 	"reflect"
 	"sort"
 )
 
-// Misc top-level stuff
-
-type vmExt struct {
-	value  string
-	isCode bool
-}
-
-type vmExtMap map[string]vmExt
-
-// RuntimeError is an error discovered during evaluation of the program
-type RuntimeError struct {
-	StackTrace []TraceFrame
-	Msg        string
-}
-
-func makeRuntimeError(loc *LocationRange, msg string) RuntimeError {
-	// TODO(dcunnin): Build proper stacktrace.
-	return RuntimeError{
-		StackTrace: []TraceFrame{
-			{
-				Loc:  *loc,
-				Name: "name",
-			},
-		},
-		Msg: msg,
-	}
-}
-
-func (err RuntimeError) Error() string {
-	// TODO(dcunnin): Include stacktrace.
-	return err.Msg
-}
-
-// Values and state
-
-type bindingFrame map[identifier]*thunk
-
-type value interface {
-}
-
-type valueString struct {
-	value string
-}
-
-func makeValueString(v string) *valueString {
-	return &valueString{value: v}
-}
-
-type valueBoolean struct {
-	value bool
-}
-
-func makeValueBoolean(v bool) *valueBoolean {
-	return &valueBoolean{value: v}
-}
-
-type valueNumber struct {
-	value float64
-}
-
-func makeValueNumber(v float64) *valueNumber {
-	return &valueNumber{value: v}
-}
-
-// TODO(dcunnin): Maybe intern values null, true, and false?
-type valueNull struct {
-}
-
-func makeValueNull() *valueNull {
-	return &valueNull{}
-}
-
-type thunk struct {
-	content  value // nil if not filled
-	name     identifier
+// TODO(sbarzowski) use it as a pointer in most places b/c it can sometimes be shared
+// for example it can be shared between array elements and function arguments
+type environment struct {
+	sb selfBinding
+
+	// Bindings introduced in this frame. The way previous bindings are treated
+	// depends on the type of a frame.
+	// If isCall == true then previous bindings are ignored (it's a clean
+	// environment with just the variables we have here).
+	// If isCall == false then if this frame doesn't contain a binding
+	// previous bindings will be used.
```

Comment:
Not sure what you mean by composed. The lookup starts at the most nested one and then we try outer ones. We finish when we reach a "call" frame.